src/Transpiler.ts

```diff
@@ -584,35 +596,80 @@ export abstract class LuaTranspiler {
     }
 
     public transpileForOf(node: ts.ForOfStatement): string {
-        // Get variable identifier
-        const variable = (node.initializer as ts.VariableDeclarationList).declarations[0];
-
         // Transpile expression
         const iterable = this.transpileExpression(node.expression);
 
-        // Use ipairs for array types, pairs otherwise
-        const isArray = tsHelper.isArrayType(this.checker.getTypeAtLocation(node.expression), this.checker);
-
         let result = "";
 
-        if (!isArray && ts.isIdentifier(variable.name)) {
-            result = this.indent + `for _, ${this.transpileIdentifier(variable.name)} in pairs(${iterable}) do\n`;
+        let itemVariable: ts.Expression;
+        if (tsHelper.isArrayType(this.checker.getTypeAtLocation(node.expression), this.checker)) {
+            // Arrays use numeric for loop (performs better than ipairs)
+            const indexVariable = `____TS_index`;
```

Comment:
Since the temp is only accessed in the next line after the for statement, any nested loops will be re-declaring the temp in their own scope which lua allows for:
```ts
declare const arr: string[];
for (let a of arr) {
    for (let b of arr) {}
}
```
```lua
for ____TS_index=1, #arr do
    local a = arr[____TS_index]; --Only place outer temp will ever be referenced
    do
        for ____TS_index=1, #arr do --Redeclares temp as a local in the inner for loop
            local b = arr[____TS_index]; --References inner temp
            do
            end
        end
    end
end
```